command/agent/retry_join.go

```diff
@@ -107,24 +170,19 @@ func (r *retryJoiner) RetryJoin(serverJoin *ServerJoin) {
 	addrsToJoin := strings.Join(serverJoin.RetryJoin, " ")
 	r.logger.Info("starting retry join", "servers", addrsToJoin)
 
-	standardLogger := r.logger.StandardLogger(&log.StandardLoggerOptions{InferLevels: true})
 	for {
 		var addrs []string
 		var n int
 		var err error
 
 		for _, addr := range serverJoin.RetryJoin {
-			switch {
-			case strings.HasPrefix(addr, "provider="):
-				servers, err := r.discover.Addrs(addr, standardLogger)
-				if err != nil {
-					r.logger.Error("determining join addresses failed", "error", err)
-				} else {
-					addrs = append(addrs, servers...)
-				}
-			default:
-				addrs = append(addrs, addr)
+			servers, err := r.autoDiscover.Addrs(addr, r.logger)
+			if err != nil {
+				r.logger.Error("discovering join addresses failed", "join_config", addr, "error", err)
+				return
```

Comment:
@tgross @acaloiaro Is this `return` that was added here necessary and intended? We have found that this makes the process quite brittle as a single failure of the discovery causes it to give up (even if infinite retries are specified). Looking at the way the code was previously, this was not the case.

Reply:
I created https://github.com/hashicorp/nomad/issues/24560 related to this.

Reply:
@ryansammonaiven  I no longer have the cotext of this pull request in my head. 

However, your conclusion appears correct after a quick review of the code. I believe I got this return wrong and `L182` should be removed so that the retry count advances in `L205`.

The test suite should probably be extended to ensure retries on failure.

Reply:
To close the loop in this thread, this is being addressed in #24561.